apps/jupyter/extensions/notebook-runner/src/notebook-runner-state.ts

```diff
@@ -0,0 +1,276 @@
+        );
+      };
+    });
+  }
+
+  private async prepareOtterKernel(kernel: IKernelConnection): Promise<void> {
+    console.log("Preparing Otter kernel:", kernel.name);
+
+    console.log("Importing basic libraries..");
+    await kernel.requestExecute(
+      {
+        code: `
+        from ipykernel.comm import Comm
+      `,
+      },
+      true,
+    ).done;
+
+    await importBasePackages(kernel);
+    await installOtter(kernel);
+    await installNbConvert(kernel);
+
+    console.log("Importing Otter Grader..");
+    await kernel.requestExecute(
+      {
+        code: `
+      # import otter
+      from otter.assign import main as assign
+      from otter.run import main as run
+      `,
+      },
+      true,
+    ).done;
+
+    console.log("Otter Grader is ready to be used!", kernel);
+    setKernelIsPrepared(kernel);
+  }
+
+  public async restartOtterKernel(): Promise<void> {
+    const sessionContext = await this.otterSessionContext;
+
+    await sessionContext.restartKernel();
+  }
+
+  public async getOtterKernel(): Promise<IKernelConnection> {
+    const sessionContext = await this.otterSessionContext;
+
+    while (!sessionContext.session?.kernel) {
+      console.debug(
+        "No kernel available in otter session context, restarting the kernel..",
+        sessionContext.session,
+      );
+      await sessionContext.startKernel();
+    }
+
+    console.debug(
+      "Kernel is available in otter session, waiting for it to be prepared:",
+      sessionContext.session.kernel,
+    );
+    await waitForKernelToBePrepared(sessionContext.session.kernel);
+
+    return sessionContext.session.kernel;
+  }
+
+  readJsonFromVirtualFilesystem = async <T>(
+    kernel: IKernelConnection,
+    path: string,
+  ): Promise<T> => {
+    let receiveResults: (results: T) => void = () => {};
+    let rejectResults: (reason?: Error) => void = () => {};
+    const waitForResults = new Promise<T>((resolve, reject) => {
+      receiveResults = resolve;
+      rejectResults = reject;
+    });
+    const resultTimeout = setTimeout(() => {
+      rejectResults(new Error(`Timeout waiting for reading ${path}`));
+    }, 1000 * 5);
+
+    const commListener: CommListener = async (comm, msg) => {
+      clearTimeout(resultTimeout);
+      this.removeCommListener(commListener);
+
+      const { data } = msg.content;
+      if (typeof data.results === "string") {
+        receiveResults(JSON.parse(data.results));
+      }
+
+      rejectResults(
+        new Error("Invalid results message: " + JSON.stringify(data)),
+      );
+    };
+
+    this.addCommListener(commListener);
+
+    await kernel.requestExecute({
+      code: `
+with open("${path}", "r", encoding="utf-8") as f:
+  json_content = f.read()
+
+comm = Comm(target_name='notebook-runner')
+comm.send(data={'results': json_content})
+`,
+    }).done;
+
+    return waitForResults;
+  };
+
+  readBinaryFromVirtualFilesystem = async (
```

Comment:
This is very similar to `readJson` , couldn't we have a single function to deal with all of the shared logic?